src/Tests/Microsoft.Diagnostics.Monitoring.UnitTestApp/Scenarios/FunctionProbes/FunctionProbesScenario.cs

```diff
@@ -0,0 +1,326 @@
+        {
+            Assert.NotNull(method);
+
+            probeProxy.RegisterPerFunctionProbe(method, (object[] actualArgs) =>
+            {
+                if (hasImplicitThis)
+                {
+                    Assert.NotEmpty(actualArgs);
+                    Assert.Equal(capturedThisObj, actualArgs[0]);
+                    Assert.Equal(args, actualArgs.Skip(1));
+                }
+                else
+                {
+                    Assert.Equal(args, actualArgs);
+                }
+            });
+
+            await WaitForProbeInstallationAsync(probeManager, probeProxy, new[] { method }, token);
+
+            await invoker().WaitAsync(token);
+
+            Assert.Equal(1, probeProxy.GetProbeInvokeCount(method));
+        }
+
+        private static async Task WaitForProbeUninstallationAsync(FunctionProbesManager probeManager, PerFunctionProbeProxy probeProxy, CancellationToken token)
+        {
+            probeManager.StopCapturing();
+
+            MethodInfo uninstallationTestMethod = typeof(FunctionProbesScenario).GetMethod(nameof(FunctionProbesScenario.UninstallationTestStub));
+            Assert.NotNull(uninstallationTestMethod);
+
+            probeProxy.RegisterPerFunctionProbe(uninstallationTestMethod, (object[] args) => { });
+
+            while (!token.IsCancellationRequested)
+            {
+                int currentCount = probeProxy.GetProbeInvokeCount(uninstallationTestMethod);
+                uninstallationTestMethod.Invoke(null, null);
+                if (currentCount == probeProxy.GetProbeInvokeCount(uninstallationTestMethod))
+                {
+                    return;
+                }
+
+                await Task.Delay(100, token);
+            }
+
+            token.ThrowIfCancellationRequested();
+        }
+
+        private static async Task WaitForProbeInstallationAsync(FunctionProbesManager probeManager, PerFunctionProbeProxy probeProxy, IList<MethodInfo> methods, CancellationToken token)
+        {
+            // Register the uninstallation test method as well so WaitForProbeUninstallationAsync can function
+            MethodInfo uninstallationTestMethod = typeof(FunctionProbesScenario).GetMethod(nameof(FunctionProbesScenario.UninstallationTestStub));
+            Assert.NotNull(uninstallationTestMethod);
+
+            MethodInfo installationTestMethod = typeof(FunctionProbesScenario).GetMethod(nameof(FunctionProbesScenario.InstallationTestStub));
+            Assert.NotNull(installationTestMethod);
+
+            probeProxy.RegisterPerFunctionProbe(installationTestMethod, (object[] args) => { });
+
+            List<MethodInfo> methodsToCapture = new(methods.Count + 2)
+            {
+                installationTestMethod,
+                uninstallationTestMethod
+            };
+            methodsToCapture.AddRange(methods);
+            probeManager.StartCapturing(methodsToCapture);
```

Comment:
Doesn't have to be done in this PR...

I'd like to see a callback/synchronization mechanism here so that way we can definitively tell when the probes have been registered. As it is, this effective tosses the methods over a fence and the caller never gets to know "were all of the methods processed?", "did some failure occur?", etc. Adding this mechanism allows the caller to choose to observe failures and to choose to wait for the probes to be inserted before continuing.

This would then alleviate the need to effectively poll for when the probes were actually inserted in this test and then you could probably remove the need for the `InstallationTestStub` and `UninstallationTestStub` methods.